File: packages/thoughtflow/src/thoughtflow/knowledge-impact.ts
type JsonObject = Record<string, any>;
type RefKey = string;

const unitFields = ["missing_evidence_node_refs", "missing_prerequisite_refs", "ref", "status"];
const UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/;

function object(value: unknown): value is JsonObject {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function compareBytes(left: string, right: string) {
  return Buffer.compare(Buffer.from(left), Buffer.from(right));
}

function issue(code: string, path: string) {
  return { code, path, severity: "error" };
}

function notEvaluated(code: string, path: string) {
  return { object_result: "not_evaluated", operation_outcome: "indeterminate", issues: [issue(code, path)], impacted_steps: [] };
}

function refKey(value: unknown): RefKey | undefined {
  if (!object(value) || Object.keys(value).sort().join(",") !== "id,revision") return undefined;
  if (typeof value.id !== "string" || !UUID.test(value.id) || !Number.isSafeInteger(value.revision) || value.revision < 1) return undefined;
  return `${value.id}\u0000${String(value.revision).padStart(16, "0")}`;
}

function refFromKey(key: RefKey) {
  const [id, revision] = key.split("\u0000");
  return { id, revision: Number(revision) };
}

function refList(value: unknown): RefKey[] | undefined {
  if (!Array.isArray(value)) return undefined;
  const keys = value.map(refKey);
  return keys.some((key) => key === undefined) ? undefined : keys as RefKey[];
}

function canonicalKeys(keys: RefKey[]) {
  const sorted = [...keys].sort(compareBytes);
  return new Set(keys).size === keys.length && keys.every((key, index) => key === sorted[index]);
}

function canonicalRefSet(value: unknown): RefKey[] | undefined {
  const keys = refList(value);
  return keys === undefined || !canonicalKeys(keys) ? undefined : keys;
}

function projectionReasons(projection: unknown): Map<RefKey, JsonObject> | undefined {
  if (!object(projection) || projection.object_result !== "valid" || projection.operation_outcome !== "succeeded" || !Array.isArray(projection.issues) || projection.issues.length !== 0 || !Array.isArray(projection.units)) return undefined;
  const reasons = new Map<RefKey, JsonObject>(), unitRefs: RefKey[] = [];
  for (const unit of projection.units) {
    if (!object(unit) || Object.keys(unit).sort().join(",") !== unitFields.join(",")) return undefined;
    const key = refKey(unit.ref), prerequisites = canonicalRefSet(unit.missing_prerequisite_refs), evidence = canonicalRefSet(unit.missing_evidence_node_refs);
    if (key === undefined || prerequisites === undefined || evidence === undefined || !["blocked", "needs_evidence", "ready"].includes(unit.status) || reasons.has(key)) return undefined;
    unitRefs.push(key);
    reasons.set(key, {
      knowledge_unit_ref: refFromKey(key), status: unit.status,
      missing_prerequisite_refs: prerequisites.map(refFromKey),
      missing_evidence_node_refs: evidence.map(refFromKey),
    });
  }
  return canonicalKeys(unitRefs) ? reasons : undefined;
}

export function projectKnowledgeImpacts(flow: unknown, projection: unknown) {
  const reasonsByRef = projectionReasons(projection);
  if (reasonsByRef === undefined) return notEvaluated("thoughtflow.knowledge_impact.invalid_projection", "/projection");
  if (!object(flow) || !Array.isArray(flow.steps)) return notEvaluated("thoughtflow.knowledge_impact.invalid_flow", "/flow/steps");

  const impacts = new Map<string, Map<RefKey, JsonObject>>();
  for (let index = 0; index < flow.steps.length; index += 1) {
    const step = flow.steps[index], path = `/flow/steps/${index}`;
    if (!object(step) || typeof step.step_id !== "string" || typeof step.kind !== "string") return notEvaluated("thoughtflow.knowledge_impact.invalid_flow", path);
    const refs = refList(step.knowledge_unit_refs);
    if (refs === undefined) return notEvaluated("thoughtflow.knowledge_impact.invalid_flow", `${path}/knowledge_unit_refs`);
    if (step.kind === "operation") {
      const behaviorRef = object(step.behavior_ref) ? refKey(step.behavior_ref.knowledge_unit_ref) : undefined;
      if (behaviorRef === undefined) return notEvaluated("thoughtflow.knowledge_impact.invalid_flow", `${path}/behavior_ref/knowledge_unit_ref`);
      refs.push(behaviorRef);
    }
    const expectedStatus = ["analysis", "operation"].includes(step.kind) ? "blocked" : step.kind === "verification" ? "needs_evidence" : undefined;
    if (expectedStatus === undefined) continue;
    for (const ref of new Set(refs)) {
      const reason = reasonsByRef.get(ref);
      if (reason?.status === expectedStatus) {
        if (!impacts.has(step.step_id)) impacts.set(step.step_id, new Map());
        impacts.get(step.step_id)!.set(ref, reason);
      }
    }
  }
  const impacted_steps = [...impacts.entries()].sort(([left], [right]) => compareBytes(left, right)).map(([step_id, byRef]) => ({
    step_id,
    reasons: [...byRef.entries()].sort(([left], [right]) => compareBytes(left, right)).map(([, reason]) => reason),
  }));
  return { object_result: "valid", operation_outcome: "succeeded", issues: [], impacted_steps };
}

File: packages/thoughtflow/python/intelliengine_thoughtflow/knowledge_impact.py
from __future__ import annotations

import re
from typing import Any


_REF_FIELDS = {"id", "revision"}
_UNIT_FIELDS = {"ref", "status", "missing_prerequisite_refs", "missing_evidence_node_refs"}
_SAFE_INTEGER = 9007199254740991
_UUID = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$")


def _issue(code: str, path: str) -> dict:
    return {"code": code, "path": path, "severity": "error"}


def _not_evaluated(code: str, path: str) -> dict:
    return {
        "object_result": "not_evaluated",
        "operation_outcome": "indeterminate",
        "issues": [_issue(code, path)],
        "impacted_steps": [],
    }


def _ref_key(value: Any) -> tuple[bytes, int] | None:
    if not isinstance(value, dict) or set(value) != _REF_FIELDS:
        return None
    identifier, revision = value.get("id"), value.get("revision")
    if not isinstance(identifier, str) or _UUID.fullmatch(identifier) is None or isinstance(revision, bool) or not isinstance(revision, int) or not 1 <= revision <= _SAFE_INTEGER:
        return None
    return identifier.encode("utf-8"), revision


def _ref_from_key(key: tuple[bytes, int]) -> dict:
    return {"id": key[0].decode("utf-8"), "revision": key[1]}


def _ref_list(value: Any) -> list[tuple[bytes, int]] | None:
    if not isinstance(value, list):
        return None
    keys = [_ref_key(item) for item in value]
    if any(key is None for key in keys):
        return None
    return [key for key in keys if key is not None]


def _projection_reasons(projection: Any) -> dict[tuple[bytes, int], dict] | None:
    if not isinstance(projection, dict) or projection.get("object_result") != "valid" or not isinstance(projection.get("units"), list):
        return None
    reasons: dict[tuple[bytes, int], dict] = {}
    for unit in projection["units"]:
        if not isinstance(unit, dict) or set(unit) != _UNIT_FIELDS:
            return None
        key = _ref_key(unit.get("ref"))
        prerequisites = _ref_list(unit.get("missing_prerequisite_refs"))
        evidence = _ref_list(unit.get("missing_evidence_node_refs"))
        if key is None or prerequisites is None or evidence is None or unit.get("status") not in {"blocked", "needs_evidence", "ready"} or key in reasons:
            return None
        reasons[key] = {
            "knowledge_unit_ref": _ref_from_key(key),
            "status": unit["status"],
            "missing_prerequisite_refs": [_ref_from_key(ref) for ref in prerequisites],
            "missing_evidence_node_refs": [_ref_from_key(ref) for ref in evidence],
        }
    return reasons


def project_knowledge_impacts(flow: Any, projection: Any) -> dict:
    reasons_by_ref = _projection_reasons(projection)
    if reasons_by_ref is None:
        return _not_evaluated("thoughtflow.knowledge_impact.invalid_projection", "/projection")
    if not isinstance(flow, dict) or not isinstance(flow.get("steps"), list):
        return _not_evaluated("thoughtflow.knowledge_impact.invalid_flow", "/flow/steps")

    impacts: dict[str, dict[tuple[bytes, int], dict]] = {}
    for index, step in enumerate(flow["steps"]):
        path = f"/flow/steps/{index}"
        if not isinstance(step, dict) or not isinstance(step.get("step_id"), str) or not isinstance(step.get("kind"), str):
            return _not_evaluated("thoughtflow.knowledge_impact.invalid_flow", path)
        refs = _ref_list(step.get("knowledge_unit_refs"))
        if refs is None:
            return _not_evaluated("thoughtflow.knowledge_impact.invalid_flow", f"{path}/knowledge_unit_refs")
        if step["kind"] == "operation":
            behavior = step.get("behavior_ref")
            behavior_ref = _ref_key(behavior.get("knowledge_unit_ref")) if isinstance(behavior, dict) else None
            if behavior_ref is None:
                return _not_evaluated("thoughtflow.knowledge_impact.invalid_flow", f"{path}/behavior_ref/knowledge_unit_ref")
            refs.append(behavior_ref)
        expected_status = "blocked" if step["kind"] in {"analysis", "operation"} else "needs_evidence" if step["kind"] == "verification" else None
        if expected_status is None:
            continue
        for ref in set(refs):
            reason = reasons_by_ref.get(ref)
            if reason is not None and reason["status"] == expected_status:
                impacts.setdefault(step["step_id"], {})[ref] = reason

    impacted_steps = [
        {"step_id": step_id, "reasons": [by_ref[ref] for ref in sorted(by_ref)]}
        for step_id, by_ref in sorted(impacts.items(), key=lambda item: item[0].encode("utf-8"))
    ]
    return {
        "object_result": "valid",
        "operation_outcome": "succeeded",
        "issues": [],
        "impacted_steps": impacted_steps,
    }